File: Multiple Linear Regresion/ToyotaCorolla.R
corolla <- read_csv(file.choose())
View(corolla)

corolla <- corolla[c("Price","Age_08_04","KM","HP","cc","Doors","Gears","Quarterly_Tax","Weight")]
View(corolla)

# Data Preparation & EDA

# checking for missing values

sum(is.na(corolla)) # No missing values

# 6 - point Summary

summary(corolla)
colnames(corolla)

attach(corolla)

library(dplyr)
summarise_if(corolla , is.numeric, sd) # Standard Deviation
summarise_if(corolla , is.numeric, max) - summarise_if(corolla , is.numeric, min) # Range

# Skewness
#install.packages("moments")
library(moments)
skewness(corolla)

#Kurtosis
kurtosis(corolla)

#Graphical Representation

#Distribution
dens <- apply(corolla , 2 , density)
plot(NA, xlim=range(sapply(dens, "[", "x")), ylim=range(sapply(dens, "[", "y")))
mapply(lines, dens, col=1:length(dens))
legend("topright", legend=names(dens), fill=1:length(dens))

#Box Plot
boxplot(corolla)

# Pair plot
pairs(corolla)

# Correlation Matrix
panel.cor <- function(x, y, digits=2, prefix="", cex.cor)
{
  #usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r = (cor(x, y))
  txt <- format(c(r, 0.123456789), digits=digits)[1]
  txt <- paste(prefix, txt, sep="")
  if(missing(cex.cor)) cex <- 0.4/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex)
}
pairs(corolla, upper.panel=panel.cor,main="Scatter Plot Matrix with Correlation Coefficients")


#Linear Model
model1 <- lm(Price ~ . , data = corolla)
summary(model1)

# Output Summary:
# 1. p-values of most of the coeffs are less than 0.05 but cc and Doors are insignificant
# R squared adj and R squared are closeby
# R squared value is 0.864 - Determination is strong 


# Diagnostic Plots
#install.packages("car")
library(car)
plot(model1)


# Deletion Diagnostics for identifying influential variable
influence.measures(model1)
influenceIndexPlot(model1, id.n=3) 
influencePlot(model1, id.n=3) # Points 81, 222 & 961 seems to be outliers

#Linear Model excluding 81, 222 & 961
model2 <- lm(Price ~ . , data=corolla[-c(81 , 222 , 961),])
summary(model2)

# Model Summary
# 1. p-values the coeffs are less than 0.05
# R squared adj and R squared are closeby
# R squared value is 0.885 - Determination is strong ; better than model 1

# Variance Inflation Factors
vif(model2)  # No collinearity

# Added Variable Plots 
avPlots(model2, id.n=2, id.cex=0.8, col="red")

#Linear Model excluding 81, 222 & 961
model3 <- lm(Price ~ Age_08_04 + KM + HP + log(cc) + log(Doors) + Gears + Quarterly_Tax + Weight , data=corolla[-c(81 , 222 , 961),])
summary(model3)

# Model Summary
# 1. p-values the coeffs are less than 0.05
# R squared adj and R squared are closeby
# R squared value is 0.8857 - Determination is strong ; better than model 1


# Spliting data into train and test
n=nrow(corolla)
n1=n*0.7
n2=n-n1
train=sample(1:n,n1)
test = corolla[-train,]

pred=predict(model2,newdat=test)
actual=test$Price
error=actual-pred

test.rmse=sqrt(mean(error**2))
test.rmse

train.rmse = sqrt(mean(model2$residuals**2))
train.rmse 

File: Multiple Linear Regresion/ToyotaCorolla.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

corolla = corolla[["Price","Age_08_04","KM","HP","cc","Doors","Gears","Quarterly_Tax","Weight"]]

# Check for missing values
corolla.isna().sum()

# Basic Statistics
corolla.describe()

corolla.dtypes

#Skeweness
corolla.skew() # CC value is skewed positively

#Kurtosis
corolla.kurt() # CC value is leptokurtic

#Graphical Representation

len(corolla.columns)

i = 1
for cols in corolla.columns:
    plt.subplot(3,3,i)
    sns.distplot(corolla[cols] , hist = False , label = cols)    
    i = i+1

# Positively Skewed - Price, KM, HP, CC, Gears, Weight
# Negatively Skewed - Age
# ultiple Disributions - Doors, Quarterly_Tax


corolla.boxplot(grid = False)

# Outliers exist is all features except for Doors

# Multiple Regression Models
import statsmodels.formula.api as smf # for regression model

ml1 = smf.ols('Price ~ Age_08_04 + KM + HP + cc + Doors + Gears + Quarterly_Tax + Weight',data=corolla).fit() # regression model
ml1.summary()


# Output Summary:
# 1. p-values of most of the coeffs are less than 0.05 but cc and Doors are insignificant
# R squared adj and R squared are closeby
# R squared value is 0.864 - Determination is strong 
# AIC value is 2477
# Durbin-Watson is in the normal range

corolla.columns

# Variance Inflation Factors (VIF) for Continuous variables
rsq_age = smf.ols('Age_08_04 ~ KM + HP + cc + Doors + Gears + Quarterly_Tax + Weight' , data = corolla).fit().rsquared
vif_age = 1 / (1-rsq_age)

rsq_km = smf.ols('KM ~ Age_08_04 + HP + cc + Doors + Gears + Quarterly_Tax + Weight' , data = corolla).fit().rsquared
vif_km = 1 / (1-rsq_km)

rsq_HP = smf.ols('HP ~ Age_08_04 + KM + cc + Doors + Gears + Quarterly_Tax + Weight' , data = corolla).fit().rsquared
vif_HP = 1 / (1-rsq_HP)

rsq_cc = smf.ols('cc ~ Age_08_04 + HP + KM + Doors + Gears + Quarterly_Tax + Weight' , data = corolla).fit().rsquared
vif_cc = 1 / (1-rsq_cc)

rsq_Doors = smf.ols('Doors ~ Age_08_04 + HP + cc + KM + Gears + Quarterly_Tax + Weight' , data = corolla).fit().rsquared
vif_Doors = 1 / (1-rsq_Doors)

rsq_Gears = smf.ols('Gears ~ Age_08_04 + HP + cc + Doors + KM + Quarterly_Tax + Weight' , data = corolla).fit().rsquared
vif_Gears = 1 / (1-rsq_Gears)

rsq_tax = smf.ols('Quarterly_Tax ~ Age_08_04 + HP + cc + Doors + Gears + KM + Weight' , data = corolla).fit().rsquared
vif_tax = 1 / (1-rsq_tax)

rsq_Weight = smf.ols('Weight ~ Age_08_04 + HP + cc + Doors + Gears + Quarterly_Tax + KM' , data = corolla).fit().rsquared
vif_Weight = 1 / (1-rsq_Weight)

corolla.columns

VIF = pd.DataFrame({'variables' : ['Age_08_04', 'KM', 'HP', 'cc', 'Doors', 'Gears', 'Quarterly_Tax', 'Weight'] , 
                    'vif' : [vif_age, vif_km, vif_HP, vif_cc, vif_Doors, vif_Gears,
       vif_tax, vif_Weight]}) 

VIF # No VIF values are above 10; There is not multi collinearity

# Transforming cc and Doors
ml2 = smf.ols('Price ~ Age_08_04 + KM + HP + np.log(cc) + np.log(Doors) + Gears + Quarterly_Tax + Weight',data=corolla).fit() # regression model
ml2.summary()

# Output Summary:
# 1. p-values of most of the coeffs are less than 0.05 but Doors is insignificant; Transformation worked for cc
# R squared adj and R squared are closeby
# R squared value is 0.867 - Determination is strong - Better than model 1
# AIC value is 2473 lower than model 1
# Durbin-Watson is in the normal range

sns.distplot(corolla.Doors) # Looks to be a Categorical Feature


# Removing Door from the model
ml3 = smf.ols('Price ~ Age_08_04 + KM + HP + np.log(cc) + Gears + Quarterly_Tax + Weight',data=corolla).fit() # regression model
ml3.summary()

# Output Summary:
# 1. p-values of the coeffs are less than 0.05 
# R squared adj and R squared are closeby
# R squared value is 0.867 - Determination is strong - Better than model 1 and same as model 2
# AIC value is 2473 lower than model 1
# Durbin-Watson is in the normal range

# splitting the data into train and test

from sklearn.model_selection import train_test_split
corolla_train,corolla_test  = train_test_split(corolla,test_size = 0.3) # 30% test data

ml_final = smf.ols('Price ~ Age_08_04 + KM + HP + np.log(cc) + Gears + Quarterly_Tax + Weight',data=corolla_train).fit() # regression model
ml_final.summary()


train_pred = ml_final.predict(corolla_train)

train_res = train_pred - corolla_train.Price

train_rmse = np.sqrt(np.mean(train_res * train_res))

test_pred = ml_final.predict(corolla_test)
test_res = test_pred - corolla_test.Price
test_rmse = np.sqrt(np.mean(test_res * test_res))

train_rmse
test_rmse

# Train RMSE and Test RMSE are close by - Good Fitting